lib/Dialect/TritonGPU/Transforms/WSCodePartition.cpp

```diff
@@ -385,29 +383,63 @@ struct Channel {
 public:
   using Relation = std::pair<int, SmallVector<int>>;
 
-  Channel(int producer, SmallVector<int> &consumers, Operation *src,
-          Operation *dst, Value srcOperand, unsigned numBuffers)
-      : relation(producer, consumers), srcOp(src), dstOp(dst),
-        srcOperand(srcOperand), numBuffers(numBuffers) {}
+  Channel(int producer, SmallVector<int> &consumers, Operation *op,
+          unsigned operandIdx, unsigned numBuffers)
+      : relation(producer, consumers), op(op), operandIdx(operandIdx),
+        numBuffers(numBuffers) {}
 
   bool operator==(const Channel &c) {
-    return relation == c.relation && srcOp == c.srcOp && dstOp == c.dstOp;
+    return relation == c.relation && operandIdx == c.operandIdx && op == c.op;
   }
 
+  Operation *getDstOp() { return op; }
+  unsigned getDstOperandIdx() { return operandIdx; }
+  Value getSrcOperand() { return op->getOperand(operandIdx); }
+  Operation *getSrcOp() { return getSrcOperand().getDefiningOp(); }
+
   Relation relation; // producer task Id, a list of consumer task Ids
-  Operation *srcOp;
-  Operation *dstOp;
-  Value srcOperand;
+  Operation *op;
+  unsigned operandIdx;
```

Comment:
This is because the source operations can be changing. For example it could be a loop, or an if. The consumers are relatively stable.  `op` now means `dstOp`, and the class now provides setter functions to access srcOp,dstOp, srcOperand, dstOperand.